app/models/service_discovery/model_extensions.rb

```diff
@@ -22,35 +35,72 @@ def import_cluster_service_endpoint(cluster_service)
-                                                            details: { api_spec_content_type: cluster_service.specification_type })
-          return
-        end
+        return unless valid_cluster_service_spec?(cluster_service)
 
-        if (spec_content = cluster_service.specification).blank?
-          log_cluster_service_import_event(cluster_service, message: 'OAS specification is empty and cannot be imported')
-          return
-        end
-
-        api_docs_service = api_docs_services.build(name: cluster_service.name,
-                                                   body: spec_content,
-                                                   published: true,
-                                                   skip_swagger_validations: true)
+        api_docs_service = discovered_api_docs_service || build_api_doc_service(cluster_service)
+        api_docs_service.skip_swagger_validations = true
+        api_docs_service.body = cluster_service.specification_body
 
         unless api_docs_service.save
           log_cluster_service_import_event(cluster_service, message: 'Could not create ActiveDocs',
                                                             details: { errors: api_docs_service.errors.full_messages })
         end
       end
 
+      def discovered_api_docs_service
+        api_docs_services.discovered.first
+      end
+
       protected
 
+      def build_api_doc_service(cluster_service)
+        api_docs_services.build(name: cluster_service.name, published: true).tap { |object| object.discovered = true }
+      end
+
+      def valid_cluster_service_spec?(cluster_service)
+        cluster_service_spec_oas?(cluster_service) && cluster_service_spec_present?(cluster_service)
+      end
+
+      def cluster_service_spec_oas?(cluster_service)
+        return true if cluster_service.specification_oas?
+        log_cluster_service_import_event(cluster_service, message: 'API specification type not supported',
+                                                          details: { api_spec_content_type: cluster_service.specification_type })
+        false
+      end
+
+      def cluster_service_spec_present?(cluster_service)
+        return true if cluster_service.specification_body.present?
+        log_cluster_service_import_event(cluster_service, message: 'OAS specification is empty and cannot be imported')
+        false
+      end
+
       def log_cluster_service_import_event(cluster_service, message:, details: {})
         exception = ImportClusterDefinitionsError.new(message)
         exception_details = { service_id: id, cluster_service: { self_link: cluster_service.self_link } }.merge(details)
         System::ErrorReporting.report_error(exception, exception_details)
       end
     end
+
+    module ApiDocs
+      module Service
+        extend ActiveSupport::Concern
+
+        included do
+          class_eval do
+            attr_readonly :discovered
```

Comment:
OK 👍 but the mass-assignment warning earlier was because of `attr_accessible`
https://github.com/3scale/porta/blob/0f3a0c68c978790db9f0954d4de108552f7976ef/app/models/api_docs/service.rb#L11

Reply:
Which btw is a pain and pretty unnecessary since all controllers about ApiDocs have strong params now 😄